aces/Loris
# [instrument_builder] REDCap 2 LINST

## Brief summary of changes

This PR replaces #7528 and mirrors what dave's PR (#8181) tools/redcap2linst.php tool does but just in js and on the browser, with the exception of 2 commits which don't translate to the front-end feature:

- Add --trim-formname option to script
- Add support for using REDCap API instead of CSV file

**Additional feature**: I have added XIN rules functionality for the branching and required fields columns as an extra if its desirable (as a PR to this PR branch) https://github.com/zaliqarosli/Loris/pull/45/


#### Testing instructions (if applicable)

*Note: test on both Firefox and Chrome

1. Load a test .linst file as you would normally. Notice how the instrument is rendered on the 'Build' tab. Make some modifications to the instrument if you like. Save the instrument under the 'Save' tab. Look at the linst file and make sure it is what you expect.
2. Take a REDCap .csv data dictionary (If you don't have one, let me know and I can share one with you). Load the .csv file using the appropriate loader. All the instruments in the data dictionary should download in your browser as multiple .linst files in a zip folder. Make sure the number of instruments match. Read the .linst files and make sure it is what you expect, following the same format as the linst file generated from step 1.
3. Repeat step 1 for one of the downloaded linst files.
4. Generate the SQL file for the instrument: cat $instrument.linst | php tools/generate_tables_sql_and_testNames.php
Check the SQL statements were generated fine.
5. Run the SQL script and make sure you get no SQL errors.
6. Add the instrument to the test_battery table.
7. Run php tools/assign_missing_instruments.php
8. Test the instrument on the front end, and make sure it works like any normal linst instrument

#### Link(s) to related issue(s)

* Resolves #  (Reference the issue this fixes, if any.)
